### PR TITLE
Implement a 2-stage loading scheme for nodes.

### LINKED ISSUE #380: Circular reference with group control curves
> [‎07/‎02/‎2017 12:40] Arnott, Joshua: 
> I came across an issue with a circular reference in Pywr yesterday.
> I tried to set the cost of a reservoir to a control curve parameter which used the storage from an aggregated storage which included the original reservoir.
> It didn't like that.
> [‎07/‎02/‎2017 12:40] Tomlinson, James E: 
> I had the same thing this morning
> FYI I came across it ... and worked around it
> [‎07/‎02/‎2017 12:41] Arnott, Joshua: 
> I just put the cost on the WTW instead.
> [‎07/‎02/‎2017 12:41] Tomlinson, James E: 
> If you set it to the node directly it works
> [‎07/‎02/‎2017 12:41] Arnott, Joshua: 
> In Python-space?
> [‎07/‎02/‎2017 12:42] Tomlinson, James E: 
> In JSON
> [‎07/‎02/‎2017 12:42] Arnott, Joshua: 
> You mean not in the parameters section?
> [‎07/‎02/‎2017 12:42] Tomlinson, James E: 
> Yes

### PR DESCRIPTION
**Pull request by jetuk**
The changes here remove the existing `load` classmethod for
nodes. These are replaced with a new classmethod `pre_load`
which performs the object instantiation. After this itial
stage the parameters and recorders are fully loaded. Finally,
any parameter references are assigned to node attributes.

This resolves the circular loading between nodes and parameters.
However, it has required somworking of node init methods
which has impacted on the JSON schema for the nodes. Overall
this should make the JSON schema and init method arguments
more consistent.

Fixes #380.

### CODE REVIEW

File: pywr/_model.pyx [hunk 1]
```diff
@@ -440,20 +444,21 @@ class Model(object):
         logger.info('Model load complete!')
         return model
 
-    @classmethod
-    def _get_node_from_ref(cls, model, node_name):
```
Comment (jetuk) [inline pywr/_model.pyx:444]: Reinstate this and defer to `model.nodes`, but with a deprecation warning.

Comment (jetuk) [reply pywr/_model.pyx:444]: Done in latest commit.

File: pywr/nodes.py [hunk 2]
```diff
@@ -489,15 +468,6 @@ def check(self):
...
-    def load(cls, data, model):
-        del(data["type"])
-        nodes = []
-        for node_name in data.pop("nodes"):
-            nodes.append(model._get_node_from_ref(model, node_name))
-        node = cls(model, nodes=nodes, **data)
-        return node
-
 
 class RollingVirtualStorage(Drawable, _core.RollingVirtualStorage, metaclass=NodeMeta):
```
Comment (Batch21) [inline pywr/nodes.py:472]: This need to have the `Loadable`  Mixin added, as well as these attributes:

```python
 __parameter_attributes__ = ('min_volume', 'max_volume')
__node_attributes__ = ('nodes', )
```

File: pywr/nodes.py [hunk 3]
```diff
@@ -884,16 +867,6 @@ def iter_slots(self, slot_name=None, is_connector=True):
...
-        # max_flow and cost should be lists of parameter definitions
-        max_flow = [load_parameter(model, p) for p in data.pop('max_flow')]
-        cost = [load_parameter(model, p) for p in data.pop('cost')]
-        factors = AggregatedNode.load_factors(model, data)
-
-        del(data["type"])
-        return cls(model, max_flow=max_flow, cost=cost, factors=factors, **data)
-
 
 class AggregatedStorage(Drawable, _core.AggregatedStorage, metaclass=NodeMeta):
```
Comment (Batch21) [inline pywr/nodes.py:871]: Needs to inherit `Loadable` and have attributes added.

**Review by Batch21 — CHANGES_REQUESTED** [review]
I've tested this on a few relatively complex models and generally it is working well

**Comment by jetuk** [comment]
@Batch21 I believe I have addressed your comments here.

**Review by Batch21 — APPROVED** [review]
changes look good, thanks

**Comment by jetuk** [comment]
Thanks @Batch21

**Pull request merged**